presto-main/src/main/java/com/facebook/presto/cost/FragmentStatsProvider.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.cost;
+
+import com.facebook.presto.spi.QueryId;
+import com.facebook.presto.sql.planner.plan.PlanFragmentId;
+import com.google.inject.Inject;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.IntStream;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Provides stats for completed plan fragments.
+ */
+public class FragmentStatsProvider
+{
+    private final Map<QueryFragmentIdentifier, PlanNodeStatsEstimate> fragmentStatsMap = new ConcurrentHashMap<>();
```

Comment:
Do we need to cap the memory usage for this object? I assume `invalidateStats` should be hooked to some event listener or using some cache-based map to avoid leaking queries.

Reply:
I can make this a cache to be safe instead of a ConcurrentHashMap.  That way the memory will be capped even if invalidateStats has a leak.